### PR TITLE
Assembly language

### PR DESCRIPTION
**Pull request by Th3G33k**

### CODE REVIEW

File: assembly.html.markdown [hunk 1]
```diff
@@ -0,0 +1,218 @@
+---
+language: assembly
+filename: learn.asm
+contributors:
+    - ["", ""]
```
Comment (vendethiel) [inline assembly.html.markdown]: ?

File: assembly.html.markdown [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+---
+
+Assembly language is a low-level programming language for a computer
```
Comment (vendethiel) [inline assembly.html.markdown]: which asm is that? :P

**Comment by Th3G33k** [comment]
x86 asm, intel syntax (nasm)

File: assembly.html.markdown (hunk 2 above)
Comment (Th3G33k) [inline assembly.html.markdown]: x86 asm, intel syntax

File: assembly.html.markdown [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+```asm
+; Comments
```
Comment (geoffliu) [inline assembly.html.markdown]: A little more verbose here. As a novice, I'm not sure if this is special syntax, whether "Comments" has special meaning like "section" does. Maybe "Anything from semicolon until the end are comments"?

File: assembly.html.markdown [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+section .data
+; it contains data, constants
```
Comment (geoffliu) [inline assembly.html.markdown]: `s/it/It/`. Also could you give a few examples of C code that would generate this section? (String literals, etc.)

File: assembly.html.markdown [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+doubleword_ dw 0 ; 4 bytes
+; string zero terminated
+msg db "hello, world", 0
+; array of short int
+array dw 0, 1, 2, 3, 4
+
+section .text
+; it contains all the instructions to be executed
+
+; where the program execution begins
```
Comment (geoffliu) [inline assembly.html.markdown]: It might be helpful to mention how `_start` relates to `main` in C.

File: assembly.html.markdown (hunk 2 above)
Comment (geoffliu) [inline assembly.html.markdown]: Also, are any parts of this guide 32-bit specific?

File: assembly.html.markdown [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+```asm
+; Comments
+
+section .data
+; it contains data, constants
+
+; (variable-name) define-directive initial-value
+byte_ db 0 ; 1 byte
+word_ dw 0 ; 2 bytes
+doubleword_ dw 0 ; 4 bytes
```
Comment (geoffliu) [inline assembly.html.markdown]: `s/dw/dd/`

In general, this section is pretty confusing. Could you add a little more explanation about what each part of the line means, what `db`, `dw`, `dd` mean, and that it's OK to specify multiple values, and they all get turned into an array. Also why `()` around variable name.

Side note, I don't see a `.rodata` section, but I think it's worth mentioning.

File: assembly.html.markdown [hunk 7]
```diff
@@ -0,0 +1,218 @@
+---
+language: assembly
+filename: learn.asm
```
Comment (geoffliu) [inline assembly.html.markdown]: Isn't `.s` the usual extension for assembly on Linux?

File: assembly.html.markdown [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+movsb ; movsb, movsw, movsd, [edi] = [esi]
+lodsb ; eax = [esi]
+scasb ; cmp eax, [edi]
+cmpsb ; cmp [esi], [edi]
+stosb ; [edi] = eax
+
+; CDECL calling convention
+
+; _fct (2, 3)
+push 3 ; pass arguments right to left
```
Comment (geoffliu) [inline assembly.html.markdown]: Would be nice to mention the benefits of going right to left, var-arg functions, for example.

File: assembly.html.markdown [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+; 1 exit(int error_code)
+; 3 read(uint fd, char *buf, int count)
+; 4 write(uint fd, char *buf, int count)
+; 5 open(char *filename, int flags, int mode)
+; 6 close(uint fd)
+; 8 creat(char *pathname, int mode)
+; 13 time(int *tloc)
+; 45 sys_brk(int brk) allocate memory
+
+mov eax, 4 ; write
```
Comment (geoffliu) [inline assembly.html.markdown]: Maybe explain the calling convention here. It's quite different from the C-convention, after all.

File: assembly.html.markdown [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+mov ebx, 1 ; stdout
+mov ecx, msg
+mov edx, 13
+int 80h
```
Comment (geoffliu) [inline assembly.html.markdown]: Explain what `int 80` actually means. Most programmers aren't even aware of the user-space-kernel-space divide, let alone what interrupts mean.

File: assembly.html.markdown [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+neg eax ; eax = -eax
+
+; LOGICAL
+
+and eax, ebx ; eax &= ebx
+or eax, [doubleword_] ; eax |= [doubleword_]
+xor [doubleword_], eax ; [doubleword_] ^= eax
+not eax ; eax = !eax
+
+mov al, 0b11001100
```
Comment (geoffliu) [inline assembly.html.markdown]: Add a "SHIFTS" section header here.

**Comment by geoffliu** [comment]
Hey @Th3G33k, thanks for the work! With assembly, I think there needs to be a lot more words than code, since a deep understanding of the hardware, kernel, and their interactions is required, hence all the comments. Please bear with us; we want this guide to be beginner friendly. :)

**Comment by Th3G33k** [comment]
Thank you for the review.
I will change the code and I will make a new commit.

File: assembly.html.markdown (hunk 4 above)
Comment (geoffliu) [inline assembly.html.markdown]: Actually, not string literals. This would be static variables, right?

File: assembly.html.markdown (hunk 1 above)
Comment (adambard) [inline assembly.html.markdown]: To clarify, this is where your name (real or whatever) and website should go, e.g. `["Th3G33k", "https://github.com/Th3G33k"]`

**Pull request closed without merging**